kvbc/src/ror_app_state.cpp

```diff
@@ -0,0 +1,160 @@
+    assert(blockId > 0);
+
+    Sliver blockData;
+    Sliver key = keyManipulator_->genBlockDbKey(blockId);
+    auto s = objectStoreClient_->get(key, blockData);
+    if (!s.isOK()) {
+      LOG_ERROR(logger_, "objectStoreClient_->get returned " << s);
+      return false;
+    }
+
+    LOG_DEBUG(logger_, "getPrevDigestFromBlock, blockId: " << blockId);
+
+    assert(outPrevBlockDigest);
+    auto parentDigest = concord::storage::blockchain::block::getParentDigest(blockData);
+    assert(outPrevBlockDigest);
+    memcpy(outPrevBlockDigest, parentDigest, BLOCK_DIGEST_SIZE);
+    return true;
+  }
+
+  bool putBlock(uint64_t blockId, char *block, uint32_t blockSize) override {
+    Sliver key = keyManipulator_->genBlockDbKey(blockId);
+    Sliver value = Sliver::copy(block, blockSize);
+    Status s = objectStoreClient_->put(key, value);
+    if (s.isOK()) {
+      if (blockId > lastBlockId_) {
+        lastBlockId_ = blockId;
+      }
+      // when ST runs, blocks arrive in batches in reverse order. we need to
+      // keep track on the "Gap" and to close it. Only when it is closed, the
+      // last reachable block becomes the same as the last block
+      if (blockId == lastReachableBlockId_ + 1) {
+        lastReachableBlockId_ = lastBlockId_;
+      }
+    } else {
+      LOG_ERROR(logger_, "putBlock, blockId: " << blockId << ", status: " << s);
+      return false;
+    }
+
+    // if the local write fails but the Object store write was OK, we dont worry
+    // because next time ST will write the same block again and in the worst
+    // case the block in the Object store will be overwritten.
+    using namespace concord::storage;
+    using namespace concordUtils;
+
+    SetOfKeyValuePairs meta;
+    Sliver v1 = Sliver::copy((const char *)&lastBlockId_, sizeof(lastBlockId_));
+    Sliver v2 = Sliver::copy((const char *)&lastReachableBlockId_, sizeof(lastReachableBlockId_));
+    meta.insert(std::make_pair(lastBlockIdKey, v1));
+    meta.insert(std::make_pair(lastReachableBlockIdKey, v2));
+
+    s = localClient_->multiPut(meta);
+    if (!s.isOK()) {
+      LOG_FATAL(logger_, "putBlock, couldn't write local metadata, status: " << s);
+      exit(-1);
+    }
+
+    LOG_DEBUG(logger_,
+              "putBlock, blockId: " << blockId << ", status: " << s << ", lastBlockId: " << lastBlockId_
+                                    << ", lastReachableBlockId: " << lastReachableBlockId_);
+    return true;
+  }
+
+  uint64_t getLastReachableBlockNum() override { return lastReachableBlockId_; }
+  void setLastReachableBlockNum(uint64_t blockNum) override { lastReachableBlockId_ = blockNum; }
+
+  uint64_t getLastBlockNum() override { return lastBlockId_; }
+
+ private:
+  uint64_t lastBlockId_ = 0;
+  uint64_t lastReachableBlockId_ = 0;
+
+  Sliver lastBlockIdKey = Sliver::copy(kLastBlockIdKey.c_str(), kLastBlockIdKey.size());
+  Sliver lastReachableBlockIdKey = Sliver::copy(kLastReachableBlockIdKey.c_str(), kLastReachableBlockIdKey.size());
```

Comment:
Can these be made `static`? Rationale is that they have the same value for all objects.

Reply:
I guess it doesn't really matter much as there won't be many `RoRAppStateImpl` objects.